### ISSUE TITLE
Atom selection questions

### ISSUE THREAD

**Issue by stefdoerr** [molstar/molstar#249]
Hi all,
I have a question. We want to apply representations to specific water atoms. We currently select the atoms like this
```js
atoms: (atomIds) => {
                return query.struct.generator.atomGroups({
                    'atom-test': query.core.set.has([
                        query.core.type.set(atomIds),
                        query.ammp('id')
                    ])
                });
            },
```
However this ends up slowing down our browser a lot (freezes the tab on my computer).

We compared it to this which ends up being much much faster even if we select the exact same atoms.
```js
water: () => molstar.SelectionQueries.water.expression,
```

Is there any reason why index atom selections are so much slower than expressions? I would have assumed they would be faster since molstar would not have to search for what is water.

Also one more question. Is molstar 1-indexing atoms? I initially selected atoms with 0-based index but it always missed the last atom, then I switched to 1-based indexing, but then we found some sdf file where the 1-based index doesn't work either.

**Comment by arose** [comment]
> However this ends up slowing down our browser a lot (freezes the tab on my computer).

I wonder if it is the `query.core.type.set(atomIds)` part that takes so long, that is, building the set of atom ids (@dsehnal is the set recreated for every invocation of .has?) How big is `atomIds` in your example that freezes your Browser tab? Can you share a minimal example code?


> ```ts
> water: () => molstar.SelectionQueries.water.expression
> ```

That is defined as follows

```ts
MS.struct.modifier.union([
    MS.struct.generator.atomGroups({
        'entity-test': MS.core.rel.eq([MS.ammp('entityType'), 'water'])
    })
]);
```

Due to way Structure objects are constructed we know that water atoms are always grouped in entities with no other atoms. Hence the query only needs to test one atom for each entity.

> Also one more question. Is molstar 1-indexing atoms?

Molstar does not have an atom index that is exposed in the query language which is related to the position of an atom in a file. There is the atom.id property which gets populated by atom ids read from the source file. Not all files formats have such a field. In coreCif, xyz, gro, psf and mol files the id field is created as a 0-based serial number.

**Comment by dsehnal** [comment]
In theory, there could also be a regression in the structure builder somewhere.

Could you please share your structure and the atom set? Or at least how big the sets are so that we can profile it.

**Comment by stefdoerr** [comment]
Yes, I will try to make a minimal example today to share with you. 

As for the indexing, what is the reasoning for not allowing atom index selections? They are the most "unique" identifiers in a way since otherwise you would need to supply for example [atomname, residuename, insertion, chain, segid] combinations to uniquely select an atom from a PDB file.
My issue with the lack of indexes is that I was planning on doing the atom selections on the backend and then just send to the frontend an array of atom indexes and which representation to attach to those indexes, so that we don't have to rely on a second atom selection language.

**Comment by dsehnal** [comment]
It is possible to get the index from the input file in most cases (e.g., if you don't build the structure by hand) by using ``MS.struct.atomProperty.core.sourceIndex()``. Mol* internally sorts the input data and the ``sourceIndex`` property wasn't available until later on. If you have control over the input files, using the source index property should be safe.

**Comment by stefdoerr** [comment]
[selection_standalone.zip](https://github.com/molstar/molstar/files/6968334/selection_standalone.zip)
Here is a standalone example anyway. If you comment out the water() call and use the atom() call you will see how it hangs the browser.
You can run the example with `python -m http.server` and open the html file. The code is in `molstar.init.js`

**Comment by dsehnal** [comment]
The performance issue is fixed in 2.2.2.

Using ``sourceIndex()`` vs ``ammp('id')``:

- ``ammp('id')`` (which is shorthand for ``struct.atomProperty.macromolecular.id()``) is the value of the ``_atom_site.id`` column in mmCIF (the 2nd column in the PDB file). This is usually 1-based index of the atom in the file)
- ``struct.atomProperty.core.sourceIndex()`` is zero-based index of the ATOM/HETATM record as it appears in the input file.
